lib/logstash/outputs/elasticsearch/common.rb

```diff
@@ -16,15 +16,16 @@ module Common
     VERSION_TYPES_PERMITTING_CONFLICT = ["external", "external_gt", "external_gte"]
 
     def register
+      @template_installed = Concurrent::AtomicBoolean.new(false)
       @stopping = Concurrent::AtomicBoolean.new(false)
       # To support BWC, we check if DLQ exists in core (< 5.4). If it doesn't, we use nil to resort to previous behavior.
       @dlq_writer = dlq_enabled? ? execution_context.dlq_writer : nil
 
       setup_hosts # properly sets @hosts
       build_client
-
-      install_template
+      after_successful_connection { install_template }
```

Comment:
since we use `after_successful_connection` only once, could we avoid threading here and just have a `wait_for_successful_connection` method that performs the waiting for a connection stuff in the thread calling register?:

```ruby
def wait_for_successful_connection
  sleep_interval = @retry_initial_interval
  until successful_connection? || @stopping.true?
    @logger.debug("Waiting for connectivity to Elasticsearch cluster")
    Stud.stoppable_sleep(sleep_interval) { @stopping.true? }
    sleep_interval = next_sleep_interval(sleep_interval)
  end
end
```

```ruby
def register
  #...
  setup_hosts
  build_client
  wait_for_succesful_connection
  install_template
  # ..
end
```

Reply:
I started here, but the problem with this approach is that it stops a clean shutdown if Elasticsearch never becomes available.

Reply:
why would it prevent a clean shutdown? the `wait_for_successful_connection` method would loop with the `@stopping.true?` guard + stoppable_sleep

Reply:
The `close` method of the plugin (which calls `@stopping.make_true`) never gets called if the `register` method is blocked.